### ISSUE TITLE
Accessibility - Inconsistent Behaviour adding palette items - Note is deselected when tabbing to Braille Panel

### ISSUE THREAD

**Issue by jmwmusic** [musescore/MuseScore#19511]
### Issue type

UI bug

### Bug description

Working with NVDA screen reader many palette items won't add first time.  Sometimes hitting Esc and starting over a second time they enter, sometimes going into or out of note entry mode then trying again works.  Can take 3 or 4 goes to enter a dynamic, etc.

### Steps to reproduce

With focus on a note, Tab over to the palette's menu, Arrow down to dynamics and press right arrow to expand, press space to open.  Arrow down to a dynamic mark and press enter.  About 50% of the time I get no response.  If nothing happens press Esc and repeat.  If the dynamic mark entered correctly move to another note and try entering the same dynamic again.  If this still has not worked press Esc and then N to go into Note Entry mode and try adding again.

### MuseScore Version

4

### Regression

I don't know

### Operating system

Windows 10

**Comment by zacjansheski** [comment]
Voiceover on MacOS always calls out the added dynamic, but the focus is different when adding while in note input mode which may be the focal point of this issue.

**Comment by msviolafangirl** [comment]
Hello, I also use NVDA with MuseScore, and I can confirm this is an issue. I think the real problem is that when you're on a palette item and want to select it (the screen reader always says "item not selected") that there is no feedback of any kind whether you press space or enter on it. You don't know if you've entered the palette item until you go back to normal or note input to check it, and for some reason, more often than not, the item is not entered.

**Comment by jmwmusic** [comment]
Definitely getting a different result to you.
NVDA is reading out the same order of items when I arrow onto a note whether in normal mode or note entry mode.  both give me letter, octave, english then american value, bar/beat, then dynamic, then any other signs like tempo mark, slur etc.
Not sure what you mean about entering it without a note selected as with keyboard I don't think it is possible to enter a palette item without a note selected.
Just tried entering a couple more again and yes it says the dynamic mark followed by the words not selected, but nothing happens when I hit enter.  then either using Shift Tab or Esc to return to the score and nothing has been entered.
If you can't repeat this I can ask around on some blind music lists to see if I can find any commonality between others who are having or not having this issue.
Could it have anything to do with Braille extender Add on being installed, or the Live Braille window, or something else we might have that is different to most sighted users who test with NVDA?

**Comment by MarcSabatella** [comment]
I've never encountered any problem adding palette items by keyboard, but I don't it all that often.  But I wasn't able to reproduce an issue in a dozen tries just now.

My best guess is that somehow you are actually losing the selection before entering the palette and that's why the dynamic isn't always entered.  I guess the best way to tell would be to make a screen recording so we can see and hear what you are.  Assuming you're able to reproduce the problem, of course.

**Comment by msviolafangirl** [comment]
I think Marc is correct in saying that the reason the dynamic might not always be entered is because nothing is actually selected. I think what would be useful is that when you're trying to enter a palette item with no selection, there should be feedback saying something like "nothing is selected, please select something to enter this item". Right now, there is no feedback when something hasn't been entered properly except for escaping to check the end result. If the dynamic does get entered correctly, pressing enter will result in you ending back up in note input or normal mode.

**Comment by MarcSabatella** [comment]
This could also explain why some people have reported they have better results adding palette items while in note input mode - because in note input, it's next to impossible something *not* to be selected (can happen when switching voices I guess).

**Comment by jmwmusic** [comment]
OK, I think I've worked out how the selection is being lost. It is a shame my hardware won't support being able to create a screen recording with the NVDA voice as I managed to repeat this several ways with it not working and working to prove it.  so hopefully you can all repeat it your end from my description.

The selection is being lost if you tab over the open Braille panel to get to the palettes.

1.  Go to preferences & enable the live Braille panel if you don't already have it active.
2. Make sure a note is selected in the score.
3. Press Tab 4 times to move to the Palette selection.  Open dynamics and press enter on a dynamic.
4. Nothing happens.
5. Press Esc to return to the score and check nothing has been entered.

Repeat the above but this time use Shift Tab to move backwards to get to the palettes so you aren't crossing over the open Braille panel.  The dynamic enters correctly.

Repeat the above again this time using Shift F6 to jump over the Braille panel to get to the Palettes.  (This by the way takes more keyboard strokes than just using Tab).  The dynamic enters successfully.

Go to Preferences and disable the Live Braille panel.  The initial sequence of just pressing Tab 3 times to get to the palettes now results in the dynamic entering correctly.

in between each of these and at the end I repeated the first sequence just to check it was still not working which it did not once.

So, is it possible to reorder where that live panel is in the Tab sequence, or even better the palettes.  In MuseScore 3 I could just shift tab back to the palette area which was much better, but if they can't be back immediately before the score in the sequence of Tabs could the live Braille panel shift before the score so you shift tab to get to it.  Or is there some way of making sure selection is maintained or reinstated as you tab off the Braille window onto the Palettes.

**Comment by zacjansheski** [comment]
Got it!  Tabbing to Braille window deselects the note.  Thank you!

**Comment by zacjansheski** [comment]
Tabbing from a selected note in normal mode to the braille panel deselects the note.
Does not seem to occur when a rest is selected or in note-input mode.  (Screen capture of issue on MacOS with VoiceOver)

https://github.com/musescore/MuseScore/assets/69917893/69117984-cf99-4fd4-934f-92463555963c

**Comment by jmwmusic** [comment]
It would also be helpful if there was a shortcut to toggle a show/hide for the Braille window after enabling it in the preferences.

**Comment by MarcSabatella** [comment]
Indeed, the braille panel should be a simple toggle in the View menu with an associated command, just like the other various optional panels (mixer, selection filter, etc).  I suggest opening a separate issue to request that.

**Comment by MarcSabatella** [comment]
Unfortunately, I can still reproduce this bug in current 4.2.1 builds, for **range** selections.  Not 100% of the time, but maybe 50%.  I'm not sure what makes it happen sometimes but not other times.  But the following seems to hold reliably:

1. Click a note
2. Press Shift+Right to create a range selection
3. Press Tab to move into braille panel

Result: the range selection is lost, reverting to just the originally-clicked note

However, if I then Shift+Tab back to the score view and repeat the Shift+Right, then tab back to the braille panel, it works.

Losing the range selection means that things like repeat text that require a measure to be selected are difficult to add.  But also - there isn't any great reason to require a range selection when adding repeat text.

**Comment by zacjansheski** [comment]
@shoogle   Here is a video showing the selection being retained sometimes and getting lost other times

https://github.com/musescore/MuseScore/assets/69917893/dc61b375-2570-4543-bb0a-63243d73205f

**Comment by jmwmusic** [comment]
This seems to be fixed for individual notes now, but randomly does not always retain the selection of a range of notes.
